# 🥇 Gold Layer — First Order Products
**LushProtein Analytics | Medallion Architecture**

Reads from `medallion/silver/` and `medallion/gold/`, extracts line items from each customer's first order, enriches with the product master, and writes to `medallion/gold/`.

| Source Tables | Gold Table | Key Transformations |
|---|---|---|
| `silver_orders`, `silver_products`, `gold_customer_orders`, `gold_customer_profiles` | `gold_first_order_products` | Filter to first order line items, normalise SKU, derive SKU prefix, cast numerics, enrich with product master & order metadata, flag missing SKUs, rename columns |

## 0. Setup & Paths

In [25]:
import pandas as pd
import numpy as np
import os
import re
from pathlib import Path
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', 50)
pd.set_option('display.max_colwidth', 80)

BASE       = Path.cwd()
SILVER_DIR = BASE / 'medallion' / 'silver'
GOLD_DIR   = BASE / 'medallion' / 'gold'
GOLD_DIR.mkdir(parents=True, exist_ok=True)

print(f"Base   : {BASE}")
print(f"Silver : {SILVER_DIR}")
print(f"Gold   : {GOLD_DIR}")

Base   : /Users/bennyteo/Desktop/Benny/MITB/Term 3/ISSS603 Applied Data Science for Customer Insights/Project/LushProtein_Project_Data_20260505
Silver : /Users/bennyteo/Desktop/Benny/MITB/Term 3/ISSS603 Applied Data Science for Customer Insights/Project/LushProtein_Project_Data_20260505/medallion/silver
Gold   : /Users/bennyteo/Desktop/Benny/MITB/Term 3/ISSS603 Applied Data Science for Customer Insights/Project/LushProtein_Project_Data_20260505/medallion/gold


---
## 1. First Order Products (`silver_orders` + `silver_products` → `gold_first_order_products`)

### Step 1 — Load data & get first order IDs per customer

In [26]:
# Load data 
co  = pd.read_parquet(GOLD_DIR / 'gold_customer_orders.parquet')
df  = pd.read_parquet(SILVER_DIR / 'silver_orders.parquet')
pm  = pd.read_parquet(SILVER_DIR / 'silver_products.parquet')

# Get first order IDs per customer
first_orders = co[co['is_first_order']][['customer_id', 'order_id',
                                          'processed_at', 'country_code',
                                          'channel']].copy()

# Merge repeat purchase flags from gold_customer_profiles
cp = pd.read_parquet(GOLD_DIR / 'gold_customer_profiles.parquet')
first_orders = first_orders.merge(
    cp[['customer_id', 'repeat_purchase_90d', 'days_to_second_order']],
    on='customer_id', how='left'
)

### Step 2 — Filter `silver_orders` to first order line items only

In [27]:
# Filter silver_orders to first order line items only
line_items = df[df['Line: Type'] == 'Line Item'].copy()
first_order_ids = set(first_orders['order_id'])
first_line_items = line_items[line_items['ID'].isin(first_order_ids)].copy()

print(f"First order line items: {len(first_line_items):,}")

First order line items: 26,441


### Step 3a — Normalise SKU & derive SKU prefix (product category)

In [20]:
# Normalise SKU 
first_line_items['Line: Variant SKU'] = (first_line_items['Line: Variant SKU']
                                          .str.strip().str.upper())

# Derive SKU prefix (product category) 
first_line_items['sku_prefix'] = first_line_items['Line: Variant SKU'].str.split('-').str[0]

# First match — product_master title substring match
title_to_prefix = (pm[pm['Variant SKU'].notna()]
                   .drop_duplicates(subset='Title')[['Title', 'Variant SKU']]
                   .dropna())
title_to_prefix['sku_prefix_pm'] = title_to_prefix['Variant SKU'].str.split('-').str[0]

# Fix malformed SKU prefix from product master (e.g. '724999807999_40' → 'CRE')
title_to_prefix['sku_prefix_pm'] = title_to_prefix['sku_prefix_pm'].replace(
    {'724999807999_40': 'CRE'}
)

# Sort longest title first to avoid substring conflicts
title_to_prefix = title_to_prefix.sort_values(
    by='Title', key=lambda x: x.str.len(), ascending=False
).reset_index(drop=True)

def match_pm_title(line_title):
    if pd.isna(line_title): return None
    lt = line_title.lower().strip()
    for _, row in title_to_prefix.iterrows():
        if row['Title'].lower().strip() in lt:
            return row['sku_prefix_pm']
    return None

# Second match - keyword match for remaining null values
def infer_category(title):
    if pd.isna(title): return None
    title = title.lower()
    if 'better whey'  in title:                                                return 'BET'
    if 'clear'        in title:                                                return 'CLEAR'
    if 'lean'         in title:                                                return 'LEAN'
    if 'plant'        in title:                                                return 'PLT'
    if 'collagen'     in title:                                                return 'COL'
    if 'soy'          in title:                                                return 'SOY'
    if 'creatine'     in title:                                                return 'CRE'
    if 'prime whey'   in title:                                                return 'PRI'
    if 'elite'        in title:                                                return 'ELT'
    if any(x in title for x in ['shaker','bottle','hat','shirt',
                                 'band','sweater','tote','tee']):              return 'ACC'
    if any(x in title for x in ['pack','bundle','starter',
                                 'sampler','variety']):                        return 'BND'
    if any(x in title for x in ['capsule','omega','l-carnitine',
                                 'green tea','multivitamin','cla','bcaa',
                                 'pureburn','thermogenic']):                   return 'CAP'
    return None

# Apply to null-SKU rows only
null_mask = first_line_items['sku_prefix'].isna()

# Perform product_master match
first_line_items.loc[null_mask, 'sku_prefix'] = (
    first_line_items.loc[null_mask, 'Line: Title'].apply(match_pm_title)
)

# Perform keyword match
still_null = first_line_items['sku_prefix'].isna()
first_line_items.loc[still_null, 'sku_prefix'] = (
    first_line_items.loc[still_null, 'Line: Title'].apply(infer_category)
)

# Flag category source
first_line_items['category_source'] = 'sku'
first_line_items.loc[
    first_line_items['Line: Variant SKU'].isna() &
    first_line_items['Line: Title'].apply(lambda x: match_pm_title(x) is not None if pd.notna(x) else False),
    'category_source'
] = 'pm_match'
first_line_items.loc[
    (first_line_items['Line: Variant SKU'].isna()) &
    (first_line_items['category_source'] == 'sku'),
    'category_source'
] = 'inferred'

print(f"\nCategory source breakdown:")
print(first_line_items['category_source'].value_counts().to_string())
print(f"\nCategory breakdown after hybrid inference:")
print(first_line_items['sku_prefix'].value_counts(dropna=False).to_string())


Category source breakdown:
category_source
sku         9525
pm_match    8685
inferred    8231

Category breakdown after hybrid inference:
sku_prefix
BET      5793
LEAN     3626
CLEAR    3502
ACC      3215
PLT      2704
BND      1658
COL      1368
CAP      1164
CRE       873
SOY       836
None      767
PRI       754
ELT       181


### Step 3b - Flag free gift items

In [21]:
# Free gifts identified by explicit title indicators (🎁, FREE, 100% off)
# OR orders where ALL line items are zero-price (standalone giveaway orders)
# Bundle add-ons (e.g. free shaker with protein purchase) are NOT flagged
# since the customer made a real purchase decision
first_line_items['Line: Price'] = pd.to_numeric(first_line_items['Line: Price'], 
                                                  errors='coerce')

# Orders where ALL line items are zero-price
all_zero_orders = (line_items.groupby('ID')['Line: Price']
                   .max() == 0)
all_zero_order_ids = set(all_zero_orders[all_zero_orders].index)

first_line_items['is_free_gift'] = (
    # Explicit gift title indicators
    first_line_items['Line: Title'].str.contains(
        r'100% off|free gift|🎁|\[FREE\]|^FREE\s', 
        case=False, na=False, regex=True
    ) |
    # Entire order is zero-price (no paid items)
    first_line_items['ID'].isin(all_zero_order_ids)
)

n_gifts = first_line_items['is_free_gift'].sum()
print(f"Free gift line items flagged: {n_gifts:,} ({n_gifts/len(first_line_items):.1%})")
print(first_line_items[first_line_items['is_free_gift']]['Line: Title']
      .value_counts().head(10).to_string())

Free gift line items flagged: 1,199 (4.5%)
Line: Title
[FREE GIFT] LushProtein Classic Shaker         240
FREE LushProtein Classic Shaker                216
[FREE] LushProtein Classic Shaker              143
Collagen Glow - 300g Pack                       63
LushProtein Classic Shaker                      59
[Free gift] - Apparel                           31
🎁 LP CLASSIC SHAKER (100% off)                  31
Better Whey - 500g Pack, Chocolate Dinosaur     28
FREE - Classic Shaker                           27
Starter Pack                                    16


### Step 4 — Cast numeric columns

In [22]:
# Cast numeric columns
for col in ['Line: Quantity', 'Line: Discount', 'Line: Total']:
    first_line_items[col] = pd.to_numeric(first_line_items[col], errors='coerce').round(2)

### Step 5 — Enrich with product master & first order metadata

In [23]:
# Merge with product master
first_line_items = first_line_items.merge(
    pm[['Variant SKU', 'Title', 'Variant Grams']],
    left_on='Line: Variant SKU', right_on='Variant SKU', how='left'
).drop(columns=['Variant SKU'])

# Merge with first order metadata
first_line_items = first_line_items.merge(
    first_orders[['order_id', 'customer_id', 'processed_at', 'country_code',
                  'channel', 'repeat_purchase_90d', 'days_to_second_order']],
    left_on='ID', right_on='order_id', how='left'
).drop(columns=['order_id']) 

### Step 6 — Flag missing SKUs, select & rename final columns, save

In [24]:
# Flag orders with no SKU data
first_line_items['has_sku'] = first_line_items['Line: Variant SKU'].notna()
no_sku_orders = (~first_line_items['has_sku']).sum()
print(f"Line items without SKU: {no_sku_orders:,}")

# Select and rename final columns
gold_first_order_products = first_line_items[[
    'customer_id', 'ID', 'processed_at', 'country_code', 'channel',
    'Line: Variant SKU', 'sku_prefix', 'Line: Title', 'Line: Variant Title',
    'Title', 'Variant Grams',
    'Line: Quantity', 'Line: Price', 'Line: Discount', 'Line: Total',
    'has_sku', 'repeat_purchase_90d', 'days_to_second_order', 'category_source',
    'is_free_gift'
]].rename(columns={
    'ID'                  : 'order_id',
    'Line: Variant SKU'   : 'variant_sku',
    'sku_prefix'          : 'product_category',
    'Line: Title'         : 'line_title',
    'Line: Variant Title' : 'variant_title',
    'Title'               : 'product_title',
    'Variant Grams'       : 'variant_grams',
    'Line: Quantity'      : 'quantity',
    'Line: Price'         : 'unit_price',
    'Line: Discount'      : 'line_discount',
    'Line: Total'         : 'line_total',
})

print(f"\ngold_first_order_products: {gold_first_order_products.shape[0]:,} rows × {gold_first_order_products.shape[1]} cols")
print(f"Orders with SKU data: {gold_first_order_products['has_sku'].sum():,}")
print(f"\nProduct category breakdown:\n{gold_first_order_products['product_category'].value_counts(dropna=False).to_string()}")

# Variant-level breakdown (DTC/Shopee only — Lazada has no SKU data)
print(f"\nVariant SKU breakdown (has_sku only, n={gold_first_order_products['has_sku'].sum():,}):")
variant_summary = (gold_first_order_products[gold_first_order_products['has_sku']]
                   .groupby('variant_sku')
                   .agg(
                       n_customers   =('customer_id', 'count'),
                       avg_unit_price=('unit_price', 'mean'),
                   )
                   .round(2)
                   .sort_values('n_customers', ascending=False)
                   .reset_index())
variant_summary['avg_unit_price'] = variant_summary['avg_unit_price'].map('SGD {:.2f}'.format)
print(variant_summary.head(20).to_string(index=False))

gold_first_order_products.to_parquet(GOLD_DIR / 'gold_first_order_products.parquet', index=False)
print("\n✅ Saved: gold_first_order_products.parquet")

Line items without SKU: 16,916

gold_first_order_products: 26,441 rows × 20 cols
Orders with SKU data: 9,525

Product category breakdown:
product_category
BET      5793
LEAN     3626
CLEAR    3502
ACC      3215
PLT      2704
BND      1658
COL      1368
CAP      1164
CRE       873
SOY       836
None      767
PRI       754
ELT       181

Variant SKU breakdown (has_sku only, n=9,525):
        variant_sku  n_customers avg_unit_price
         ACC-SHK-V2         1172      SGD 11.66
  CLEAR-PEA-500G-V2         1168      SGD 62.59
    LEAN-THA-1KG-V1          930      SGD 54.77
  CLEAR-GRA-500G-V2          928      SGD 62.72
    COL-UNF-300G-V1          661      SGD 35.43
    LEAN-TAR-1KG-V1          488      SGD 56.39
     SOY-UNF-1KG-V1          460      SGD 32.56
    CRE-UNF-250G-V1          403      SGD 25.09
    LEAN-THA-40G-V1          375       SGD 4.41
    LEAN-TAR-40G-V1          349       SGD 4.38
    PLT-CHO-480G-V1          325      SGD 29.06
   CLEAR-PEA-25G-V2          261       

#### Overall
26,441 first order line items across 13,885 customers — on average each customer bought ~1.9 products in their first order.

#### SKU Coverage
- With SKU: 9,525 (36%) — exact SKU known, used for variant-level analysis
- Inferred via product master title match: 8,685 (33%) — category matched by comparing `Line: Title` against product master titles, sorted longest-first to prevent substring conflicts (e.g. `BETTER WHEY PROTEIN ELITE` matched before `BETTER WHEY PROTEIN`)
- Inferred via keyword fallback: 8,231 (31%) — category inferred from keywords in `Line: Title` for rows unmatched by product master
- Still unclassifiable: 767 (3%) — titles too ambiguous to categorise (likely free gifts, custom bundles, or discontinued products)

The original 64% null rate for `Line: Variant SKU` was predominantly driven by Lazada (99.6% of Lazada orders carry no SKU). The hybrid inference approach (product master match → keyword fallback) recovers 64% of null-SKU rows at the category level, improving overall category coverage to 97%.

Three substring conflicts were identified and resolved in the product master matching step:
- `BETTER WHEY PROTEIN` (BET) vs `BETTER WHEY PROTEIN ELITE` (ELT) — sorting by title length ensures ELT matches first
- `CLEAR PROTEIN` (CLEAR) vs `CLEAR PROTEIN STARTER PACK` (BND) — starter pack correctly classified as BND
- `LEAN PROTEIN` (LEAN) vs `LEAN PROTEIN STARTER PACK` (BND) — starter pack correctly classified as BND

One malformed SKU was also corrected: `Variant SKU = 724999807999_40` (a barcode entered in the SKU field) was remapped to prefix `CRE`.

The keyword `'cap'` was removed from the ACC inference rule as it incorrectly matched capsule product titles (e.g. `CLA Capsules`, `BCAA Capsules`), causing 518 CAP rows to be misclassified as ACC.

#### Free Gift Flag (`is_free_gift`)
A `is_free_gift` flag is applied to line items where `Line: Price == 0` or the title contains gift indicators (`🎁`, `100% off`, `free gift`). These are promotional giveaways — not customer purchase decisions — and should be excluded from gateway product and category analyses. Downstream notebooks should filter `is_free_gift == False` before computing repeat rates or category rankings.

#### Product Category Breakdown (all first orders, SKU + inferred)

| Category | Count | Notes |
|---|---|---|
| BET | 5,793 | Better Whey — most popular overall, dominant on Lazada MY |
| LEAN | 3,626 | Lean Protein |
| CLEAR | 3,502 | Clear Protein — most popular on DTC |
| ACC | 3,215 | Accessories (shakers, bottles, apparel) — filter `is_free_gift == False` for purchase analysis |
| PLT | 2,704 | Plant Protein |
| BND | 1,658 | Bundles |
| COL | 1,368 | Collagen |
| CAP | 1,164 | Capsules & supplements |
| CRE | 873 | Creatine |
| SOY | 836 | Soy Protein |
| PRI | 754 | Prime Whey |
| ELT | 181 | Elite Protein — increased from earlier runs after fixing BET/ELT substring conflict |
| None | 767 | Unclassifiable |

Note: Rankings differ from SKU-only view — Lazada data shifts BET to #1, reflecting its dominance in the Malaysian marketplace channel. CLEAR and LEAN lead in DTC, connecting to Q4 (Channel Heterogeneity) and Q9 (Geographic Segmentation).

#### Top 20 Variants by First Order Volume (DTC/Shopee only, `has_sku == True`)
Variant-level analysis is restricted to 9,525 rows where `has_sku = True`. Lazada orders are excluded as 99.6% carry no SKU. Use `variant_sku` for flavour and pack size breakdowns in downstream DS4/DS5 analysis.

| Variant SKU | Customers | Avg Unit Price | Notes |
|---|---|---|---|
| ACC-SHK-V2 | 1,172 | SGD 12.95 | Classic shaker — most common first order SKU, low-cost add-on |
| CLEAR-PEA-500G-V2 | 1,168 | SGD 103.33 | Clear Protein Peach 500g — top DTC protein SKU |
| LEAN-THA-1KG-V1 | 930 | SGD 87.20 | Lean Protein Thai Milk Tea 1KG |
| CLEAR-GRA-500G-V2 | 928 | SGD 100.28 | Clear Protein Grape 500g |
| COL-UNF-300G-V1 | 661 | SGD 69.52 | Collagen Glow Unflavoured 300g |
| LEAN-TAR-1KG-V1 | 488 | SGD 71.78 | Lean Protein Taro 1KG |
| SOY-UNF-1KG-V1 | 460 | SGD 61.75 | Soy Protein Isolate Unflavoured 1KG |
| CRE-UNF-250G-V1 | 403 | SGD 38.81 | Creatine Monohydrate 250g |
| LEAN-THA-40G-V1 | 375 | SGD 6.00 | Lean Protein Thai Milk Tea sachet — likely trial purchase |
| LEAN-TAR-40G-V1 | 349 | SGD 5.73 | Lean Protein Taro sachet — likely trial purchase |
| PLT-CHO-480G-V1 | 325 | SGD 48.38 | Plant Protein Chocolate 480g |
| CLEAR-PEA-25G-V2 | 261 | SGD 6.89 | Clear Protein Peach single serve |
| BND-SS-MIX6-V1 | 239 | SGD 30.45 | Starter pack bundle |
| CLEAR-GRA-25G-V2 | 232 | SGD 6.42 | Clear Protein Grape single serve |
| BET-UNF-1KG-V1 | 221 | SGD 83.44 | Better Whey Unflavoured 1KG |
| PLT-VAN-480G-V1 | 217 | SGD 48.90 | Plant Protein Vanilla 480g |
| PRI-UNF-1KG-V1 | 118 | SGD 151.16 | Prime Whey Isolate Unflavoured 1KG — highest price point |
| CAP-PBT-BOTTLE-V1 | 105 | SGD 77.58 | [Verify: PBT product type unclear] |
| PRI-CHO-1KG-V1 | 87 | SGD 146.67 | Prime Whey Isolate Chocolate 1KG |
| LEAN-TAR-40G-5PK-V1 | 83 | SGD 23.70 | Lean Protein Taro sachet 5-pack |